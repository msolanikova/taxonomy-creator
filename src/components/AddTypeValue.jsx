import React, {useEffect, useState} from 'react';
import {TextField} from "@mui/material";
import Button from "@mui/material/Button";
import { useDispatch, useSelector} from "react-redux";
import axios from "axios";
import {addValueToType} from "../reducers/valuesByType";

axios.defaults.timeout = 10000;

export default function AddTypeValue() {
    const types = useSelector((state) => state.types);
    const dispatch = useDispatch();
    const [newNodes, setNewNodes] = useState({});

   /* const fetchFakeApi = (id) => {
        return async function fetchFakeApiThunk(dispatch, getState) {
            try {
                const response = await axios.get(`http://localhost:8000/fake-api`, {
                    params: {
                        id: id
                    }
                });
                dispatch({
                    type: "FAKE_API_SUCCESSFUL",
                    data: response.data
                });
            } catch(err) {
                dispatch({
                    type: "FAKE_API_FAILED",
                    data: err
                })
            }
        }
    }*/

    const addTypeValue = (value, type) => {
        // dispatch(fetchFakeApi(`${type}-${value}`));
        dispatch(addValueToType({value, type}));
    }

    const changeNewNodes = (type, value) => {
        const newNodesCopy = {...newNodes};
        newNodesCopy[type] = value;
        setNewNodes(newNodesCopy);
    }

    const handleClick = (type) => {
        addTypeValue(newNodes[type], type);
        changeNewNodes(type, "");
    }

    const handleTextInput = (type, value) => {
        changeNewNodes(type, value);
    }

    const handleEnter = (event, type) => {
        if (event.key === 'Enter') {
            handleClick(type);
        }
    }

    useEffect(() => {
        const initialState = types.reduce((acc, curr) => {
            acc[curr] = "";
            return acc;
        }, {});
        setNewNodes(initialState);
    }, [types]);

    return (
        <div>
            <h2>Add Type Value</h2>
            {types.map(type => (
                <div key={`addNode-${type}`}>
                    <TextField
                        variant="outlined"
                        label={type}
                        key={`text-${type}`}
                        value={newNodes[type] ?? ""}
                        onChange={(ev) => handleTextInput(type, ev.target.value)}
                        onKeyDown={(ev) => handleEnter(ev, type)}/>
                    <Button
                        variant="contained"
                        key={`button-${type}`}
                        onClick={() => handleClick(type)}>Add node</Button>
                </div>
            ))}
        </div>
    );
}
