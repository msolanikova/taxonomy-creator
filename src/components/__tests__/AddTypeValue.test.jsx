/**
 * @jest-environment jsdom
 */

import {fireEvent, screen} from "@testing-library/react";
import AddTypeValue from "../AddTypeValue";
import React from 'react';
import {renderWithProviders} from "./test-utils";

describe("AddTypeValue component", () => {
    it("should render textbox for each type correctly", () => {
        const {container} = renderWithProviders(<AddTypeValue/>, {
            preloadedState: {
                types: ["type1", "type2", "type3"],
                valuesByType: {}
            }
        });

        expect(screen.getAllByRole("textbox")).toHaveLength(3);
        expect(screen.getAllByRole("button")).toHaveLength(3);
        expect(container).toMatchSnapshot();
    });

    it("should render no textbox if there is no type provided", () => {
        const {container} = renderWithProviders(<AddTypeValue/>, {
            preloadedState: {
                types: [],
                valuesByType: {}
            }
        });

        expect(screen.queryAllByRole("textbox")).toHaveLength(0);
        expect(screen.queryAllByRole("button")).toHaveLength(0);
        expect(container).toMatchSnapshot();
    });

    it("should dispatch CHANGE_TYPE_VALUES on hitting 'Add Type' button", async () => {
        const {store} = renderWithProviders(<AddTypeValue/>, {
            preloadedState: {
                types: ["type1", "type2", "type3"],
                valuesByType: {}
            }
        });

        fireEvent.change(screen.queryAllByRole("textbox")[1], {target: {value: 'typevalue2'}})
        fireEvent.click(screen.queryAllByRole("button")[1]);

        const state = store.getState();
        expect(Object.keys(state.valuesByType)).toHaveLength(1);
        expect(state.valuesByType).toHaveProperty("type2");
        expect(state.valuesByType["type2"]).toEqual(["typevalue2"]);

    });
});
