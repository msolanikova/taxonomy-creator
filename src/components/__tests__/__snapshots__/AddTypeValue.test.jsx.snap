// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTypeValue component should render no textbox if there is no type provided 1`] = `
<div>
  <div>
    <h2>
      Add Type Value
    </h2>
  </div>
</div>
`;

exports[`AddTypeValue component should render textbox for each type correctly 1`] = `
<div>
  <div>
    <h2>
      Add Type Value
    </h2>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          type1
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                type1
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add node
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r1:"
          id=":r1:-label"
        >
          type2
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                type2
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add node
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r2:"
          id=":r2:-label"
        >
          type3
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r2:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                type3
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add node
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
